File: 105.constructBinaryTreeFromPreorderAndInorder/constructBinaryTreFromPreorederAndInorderTraversal.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution{
public:
  TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
    int pSize = preorder.size();
    int iSize = inorder.size();
    if (!iSize || !pSize || iSize != pSize)
      return NULL;
    unordered_map<int, int> m(iSize);
    for (int i = 0; i < iSize; i++)
      m[inorder[i]] = i;
    return generate(preorder, 0 ,pSize - 1, inorder, 0, iSize - 1,  m);
  }
  TreeNode* generate(vector<int>& preorder, int pStart, int pEnd, vector<int>& inorder, int iStart, int iEnd, unordered_map<int, int>& m) {
    if (iStart > iEnd || pStart > pEnd)
      return NULL;
    TreeNode *root = new TreeNode(preorder[pStart]);
    int index = m[preorder[pStart]];
    TreeNode *leftChild = generate(preorder, pStart + 1, pStart + index - iStart, inorder, iStart, iStart + index - 1,  m);
    TreeNode *rightChild = generate(preorder, pStart + index - iStart + 1, pEnd, inorder, index + 1, iEnd, m);
    root->left = leftChild;
    root->right = rightChild;
    return root;
  }
  void postorder(TreeNode *root) {
    if (!root)
      return;
    TreeNode *leftChild = root->left;
    TreeNode *rightChild = root->right;
    postorder(leftChild);
    postorder(rightChild);
    cout << root->val << " ";
    delete root;
  }
};

int main()
{
  Solution s;
  int a[4] = {1,2,4,3};
  int b[4] = {4,2,1,3};
  vector<int> vec1(a,a+4);
  vector<int> vec2(b,b+4);
  TreeNode* res = s.buildTree(vec1, vec2);
  s.postorder(res);
  cout << endl;
  return 0;
}

File: 105.constructBinaryTreeFromPreorderAndInorder/constructBinaryTreFromPreorederAndInorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> valToIndex = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            valToIndex.put(inorder[i], i);
        }

        return buildHelper(preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1, valToIndex);
    }

    private TreeNode buildHelper(int[] preorder, int[] inorder, int pStart, int pEnd, int iStart, int iEnd, Map<Integer, Integer> valToIndex) {
        if (pStart > pEnd) return null;
        if (pStart == pEnd) return new TreeNode(preorder[pStart]);
        TreeNode res = new TreeNode(preorder[pStart]);
        int x = valToIndex.get(preorder[pStart]);
        int leftSize = x - iStart;
        res.left = buildHelper(preorder, inorder, pStart + 1, pStart + leftSize, iStart, x - 1, valToIndex);
        res.right = buildHelper(preorder, inorder, pStart + leftSize + 1, pEnd, x + 1, iEnd, valToIndex);
        return res;
    }
}